### PR TITLE
Add HD5 data generator

### PR DESCRIPTION
**Pull request by torbjorgen**
HD5 data generators for reading from dataset file.

### CODE REVIEW

File: src/rdds/lib/hdf5/hd5_data_generator.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
+import h5py
+from typing import Tuple, Union, List
+import logging
+
+from rdds.lib.logging import get_logger
+_LOGGER = get_logger(__package__)
+_LOGGER.setLevel(logging.INFO)
+
+
+DataSetNames = str  # A HDF5 data set name
```
Comment (henrikstranneheim) [inline src/rdds/lib/hdf5/hd5_data_generator.py]:
```suggestion
Hdf5DataSetNames = str
```

File: src/rdds/lib/hdf5/hd5_data_generator.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                 output_tensor_format: OutputTensorFormat = None,
+                 forever: bool = True):
+        """
+        :param hd5_file_path: HD5 file to generate data from
+        :param group_name: Group in HD5 file to read data from (reads all datasets)
+        :param output_tensor_format: The format (dataset names) constituting the output tensor
+        :param forever: Loop over data forever
+        """
+        self._hd5_file_path: str = hd5_file_path
+        self._hd5_file: h5py.File = h5py.File(self._hd5_file_path, 'r')
```
Comment (henrikstranneheim) [inline src/rdds/lib/hdf5/hd5_data_generator.py]:
```suggestion
        self._hd5_file = h5py.File(self._hd5_file_path, 'r')
```

File: src/rdds/lib/hdf5/hd5_data_generator.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        self._data_length = int
+
+        if not group_name:
+            group_name = list(self._hd5_file.keys())[0]
+        _LOGGER.info(f'Generating data from group \'{group_name}\'')
+        self._group_name: str = group_name
```
Comment (henrikstranneheim) [inline src/rdds/lib/hdf5/hd5_data_generator.py]:
```suggestion
        self._group_name: str = group_name if not group_name else list(self._hd5_file.keys())[0]
        _LOGGER.info(f'Generating data from group \'{self._group_name}\'')
```

File: src/rdds/lib/hdf5/hd5_data_generator.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        self._group: h5py.Group = self._hd5_file[self._group_name]
+
+        if not output_tensor_format:
+            # TODO: Deterministic output format
+            _LOGGER.warning('Output tensor format not deterministic and depends on HD5 content')
+            output_tensor_format = list(self._group.keys())
+        self._output_tensor_format: OutputTensorFormat = output_tensor_format
+        _LOGGER.info(f'OutputTensorFormat: {self._output_tensor_format}')
+
+        zeroeth_dataset = self._group[self._output_tensor_format[0]]
```
Comment (henrikstranneheim) [inline src/rdds/lib/hdf5/hd5_data_generator.py]: Add type hint

File: src/rdds/lib/hdf5/hd5_data_generator.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        if len(zeroeth_dataset.shape) > 1:
+            raise ValueError(f'Only 1D dataset shape supported at this time')
```
Comment (henrikstranneheim) [inline src/rdds/lib/hdf5/hd5_data_generator.py:52]: Add raise description to doc str

File: src/rdds/lib/hdf5/hd5_data_generator.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        self._idx: int = 0  # Current index for data generation
+
+        self._forever: bool = forever
+
+    @property
+    def data_length(self):
+        return self._data_length
+
+    def _assemble_output_vector(self) -> Tuple[Union[str, float], ...]:
+        output_vector = tuple()
```
Comment (henrikstranneheim) [inline src/rdds/lib/hdf5/hd5_data_generator.py]: Add doc str

File: src/rdds/lib/hdf5/hd5_data_generator.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+        self._idx: int = 0  # Current index for data generation
+
+        self._forever: bool = forever
+
+    @property
+    def data_length(self):
+        return self._data_length
+
+    def _assemble_output_vector(self) -> Tuple[Union[str, float], ...]:
```
Comment (henrikstranneheim) [inline src/rdds/lib/hdf5/hd5_data_generator.py]:
```suggestion
    def _assemble_output_vector(self) -> Tuple[Union[str, float], Any]
```
More explicit than Ellipsis

File: src/tests/fixtures/hdf5_data.py
Comment (henrikstranneheim) [inline src/tests/fixtures/hdf5_data.py:1]: You can rename this: `src/tests/conftest.py`

And import the fixture into each test function

File: src/tests/fixtures/hdf5_data.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import h5py
+from typing import Tuple
+import tempfile
+import os
+
+
+@pytest.fixture
+def hd5_file() -> Tuple[str, int]:
+    file: tempfile.NamedTemporaryFile = tempfile.NamedTemporaryFile(dir='/tmp', delete=False)
+    hd5_file: h5py.File = h5py.File(file.name, 'w')
```
Comment (henrikstranneheim) [inline src/tests/fixtures/hdf5_data.py]:
```suggestion
    hd5_file = h5py.File(file.name, 'w')
```

File: src/tests/hdf5/test_generator.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+def test_hd5_data_generator(hd5_file):
+    """
+    Test for iterator basic behavior
+    """
+    hd5_file_path, data_length = hd5_file
+    # GIVEN a HD5 data generator reading from HD5 file
+    hd5_data_generator: Hd5DataGenerator = Hd5DataGenerator(hd5_file_path=hd5_file_path, forever=False)
+    # WHEN iterating over data
+    data_iter = hd5_data_generator()
+    for i, tensor in enumerate(data_iter):
```
Comment (henrikstranneheim) [inline src/tests/hdf5/test_generator.py]:
```suggestion
    for index, tensor in enumerate(data_iter):
```

File: src/tests/hdf5/test_generator.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    assert i == data_length - 1
+
+    # THEN the generator should yield no more data
+    with pytest.raises(StopIteration):
+        data_iter.__next__()
+
+
+def test_hd5_data_generator_repeat(hd5_file):
+    """
+    Test for repeating data
```
Comment (henrikstranneheim) [inline src/tests/hdf5/test_generator.py]:
```suggestion
    Test for repeating data.
```

File: src/tests/hdf5/test_generator.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
+import pytest
+
+from rdds.lib.hdf5 import Hd5DataGenerator
+
+
+def test_hd5_data_generator(hd5_file):
+    """
+    Test for iterator basic behavior
```
Comment (henrikstranneheim) [inline src/tests/hdf5/test_generator.py]:
```suggestion
    Test for iterator basic behavior.
```

File: src/tests/hdf5/test_generator.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+        data_iter.__next__()
+
+
+def test_hd5_data_generator_repeat(hd5_file):
+    """
+    Test for repeating data
+    """
+    hd5_file_path, data_length = hd5_file
+    # GIVEN a HD5 data generator reading from HD5 file
+    hd5_data_generator: Hd5DataGenerator = Hd5DataGenerator(hd5_file_path=hd5_file_path, forever=True)
```
Comment (henrikstranneheim) [inline src/tests/hdf5/test_generator.py]:
```suggestion
    hd5_data_generator = Hd5DataGenerator(hd5_file_path=hd5_file_path, forever=True)
```

File: src/tests/hdf5/test_generator.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    # WHEN iterating over data
+    data_iter = hd5_data_generator()
```
Comment (henrikstranneheim) [inline src/tests/hdf5/test_generator.py]: Add type hint

File: src/tests/hdf5/test_generator.py [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    # THEN expect data to start over after epoch if 'forever' is True
+    for n_epochs in range(0, 2):
+        for i in range(0, data_length):
```
Comment (henrikstranneheim) [inline src/tests/hdf5/test_generator.py]:
```suggestion
        for index in range(0, data_length):
```

**Review by henrikstranneheim — APPROVED** [review]
👍

File: src/tests/fixtures/hdf5_data.py
Comment (torbjorgen) [reply src/tests/fixtures/hdf5_data.py:1]: Right - I will consider this going forward when adding more common fixtures. :+1:

**Comment by torbjorgen** [comment]
Updated according to suggestions by Henrik.

**Pull request merged**